parser.go

```diff
@@ -0,0 +1,191 @@
+			closingParaFound := false
+			for j, ch := range line {
+				if ch == ']' {
+					section = line[1:j]
+					section = strings.Trim(section, " ")
+					closingParaFound = true
+					if parsedMap.dictionary[section] == nil {
+						parsedMap.dictionary[section] = make(map[string]string)
+					}
+				}
+			}
+			if !closingParaFound {
+				ErrSectionNameMissingClosure = fmt.Errorf("section [%s] is missing closure paranthesis ']'", line[1:])
+				return ErrSectionNameMissingClosure
+			}
+			if parsedMap.dictionary == nil {
+				parsedMap.dictionary = make(map[string]map[string]string)
+			}
+
+		} else if line[0] == ']' {
+			ErrWrongParanthesisOrder = fmt.Errorf("wrong section paranthesis order, the section provided, %s, cannot start by ]", line)
+			return ErrWrongParanthesisOrder
+		} else if line[0] == ';' || line[0] == '#' || line[0] == ' ' || line[0] == '\n' || line[0] == '\t' {
+			continue
+		} else {
+			equalFound := false
+			for j, ch := range line {
+				if ch == '=' {
+					equalFound = true
+					key = line[0:j]
+					value = line[j+1:]
+					key = strings.Trim(key, " ")
+					value = strings.Trim(value, " ")
+					parsedMap.dictionary[section][key] = value
+					break
+				} else if ch == ']' {
+					ErrInvalidSectionName = fmt.Errorf("section name provided, %s, can't start with anything other than '['", line[:j+1])
+					return ErrInvalidSectionName
+				}
+			}
+			if !equalFound {
+				ErrMissingValueAssignment = fmt.Errorf("key provided, %s , is not assigned to a value, no '=' found", line)
+				return ErrMissingValueAssignment
+			}
+		}
+	}
+	return nil
+}
+
+// GetSectionNames provides section names of parsed file/string
+func (parsedMap *Parser) GetSectionNames() []string {
+	var sectionNames []string
+	for section := range parsedMap.dictionary {
+		sectionNames = append(sectionNames, section)
+	}
+	return sectionNames
+}
+
+// GetSections provides the dictionary/map structure
+func (parsedMap *Parser) GetSections() map[string]map[string]string {
+	return parsedMap.dictionary
+}
+
+// Get function takes 2 parameters: section and its key
+// Provides equivalent value
+// If section or key aren't found, it returns the zero values
+func (parsedMap *Parser) Get(sectionName, key string) string {
+	return parsedMap.dictionary[sectionName][key]
+}
+
+// Set Function takes 3 parameters: section, key and value
+// If section isn't already present, it makes the map first
+func (parsedMap *Parser) Set(sectionName, key, value string) {
+	if parsedMap.dictionary[sectionName] == nil {
+		parsedMap.dictionary[sectionName] = make(map[string]string)
+	}
+	parsedMap.dictionary[sectionName][key] = value
+}
+
+// ToString function returns a string structure of the ini file
+// Ignores redundant spaces
+func (parsedMap *Parser) String() string {
+	var stringVersion string
+	var sortedSections []string
+	for section := range parsedMap.dictionary {
+		sortedSections = append(sortedSections, section)
+	}
+	sort.Strings(sortedSections)
```

Comment:
you don't need to sort sections

Reply:
> you don't need to sort sections

I needed the sorting for testing this function, because they are randomized everytime, so I had to do so (credits to eng/Abdelrahman) and to standardize the string function somehow

Reply:
you need it for testing so do it in the tests, you can sort the returned data in the tests